extractdilkey.c

```diff
@@ -0,0 +1,386 @@
+
+#include "mlca2.h"
+#include "crystals-oids.h"
+#include "pqalgs.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define BUF_SIZE 8000
+
+int readFile(unsigned char *data,
+	     size_t *length,
+	     const char *filename);
+int writeFile(const unsigned char *data,
+	      size_t length,
+	      const char *filename);
+
+int main(int argc, char** argv)
+{
+	size_t sRawBytes = BUF_SIZE;
+	size_t sKeyBytes = BUF_SIZE;
+	int sRc = 0;
+	int sIdx = 0;
+    int sPubIn = 0;
+    int sPubOut = 0;
+    int sRawIn = 0;
+    int sRawOut = 0;
+	const char* sInFile = NULL;
+	const char* sOutFile = NULL;
+	int sPrintHelp = 0;
+    int sVerbose = 0;
+
+	for (sIdx = 1; sIdx < argc; sIdx ++)
+	{
+		if (strcmp(argv[sIdx], "-h") == 0)
+		{
+			sPrintHelp = 1;
+		}
+		else if (strcmp(argv[sIdx], "-k") == 0)
+		{
+			sIdx++;
+			sInFile = argv[sIdx];
+		}
+		else if (strcmp(argv[sIdx], "-o") == 0)
+		{
+			sIdx++;
+			sOutFile = argv[sIdx];
+		}
+		else if (strcmp(argv[sIdx], "-pubin") == 0)
+		{
+			sPubIn = 1;
+		}
+		else if (strcmp(argv[sIdx], "-pubout") == 0)
+		{
+			sPubOut = 1;
+		}
+		else if (strcmp(argv[sIdx], "-outraw") == 0)
+		{
+			sRawOut = 1;
+		}
+		else if (strcmp(argv[sIdx], "-inraw") == 0)
+		{
+			sRawIn = 1;
+		}
+        else if (strcmp(argv[sIdx], "-v") == 0)
+        {
+            sVerbose = 1;
+        }
+		else
+		{
+			printf("**** ERROR : Unknown parameter : %s\n", argv[sIdx]);
+			sPrintHelp = 1;
+		}
+	}
+
+
+	if (NULL == sInFile ||
+        (NULL == sOutFile && sRawOut)
+        )
+	{
+		printf("**** ERROR : Invalid input parms\n");
+		sPrintHelp = 1;
+	}
+
+	if (0 != sPrintHelp)
+	{
+		printf("\nextractdilkey -k <input key> [-pubin] [-inraw] [-o <output filename> [-outraw]]\n");
+		exit(0);
+	}
+
+	unsigned char* sRawKey = malloc(BUF_SIZE);
+	unsigned char* sKey = malloc(BUF_SIZE);
+
+	sRc = readFile(sKey,&sKeyBytes,sInFile);
+	if (0 != sRc)
+	{
+		printf("**** ERROR : Unable to read from : %s\n", sInFile);
+		sRc = 1;
+	}
+
+    do {
```

Comment:
You've reinvented goto.

Some alternatives that are cleaner:

- exit() on failure
- if(success == rc) { do next thing ; }